core/src/util/builders.cpp

```diff
@@ -44,65 +35,54 @@ JoinTypes JoinTypeFromString(const std::string& str) {
 
 void Builders::buildPolygon(const Polygon& _polygon, float _height, PolygonBuilder& _ctx) {
 
-    TESStesselator* tesselator = tessNewTess(&allocator);
-    isect2d::AABB bbox;
+    mapbox::Earcut<float, uint16_t> earcut;
 
-    if (_ctx.useTexCoords && _polygon.size() > 0 && _polygon[0].size() > 0) {
-        // initialize the axis-aligned bounding box of the polygon
-        bbox = isect2d::AABB(_polygon[0][0].x, _polygon[0][0].y, 0, 0);
-    }
+    earcut(_polygon);
 
-    // add polygon contour for every ring
-    for (auto& line : _polygon) {
-        if (_ctx.useTexCoords) {
-            for (auto& p : line) {
-                bbox.include(p.x, p.y);
-            }
+    uint16_t vertexDataOffset = _ctx.numVertices;
+
+    if (vertexDataOffset == 0) {
+        _ctx.indices = std::move(earcut.indices);
+    } else {
+        _ctx.indices.reserve(_ctx.indices.size() +  earcut.indices.size());
+        for (auto i : earcut.indices) {
```

Comment:
I thought of using std::transform here. But in any case the else branch
will not be taken atm as I moved the extrusion building to come after
tessellation :)

On Wed, Sep 2, 2015 at 11:46 PM, Matt Blair <notifications@github.com>
wrote:

> In core/src/util/builders.cpp
> <https://github.com/tangrams/tangram-es/pull/268#discussion_r38588850>:
>
> >
> > -    // add polygon contour for every ring
> > -    for (auto& line : _polygon) {
> > -        if (_ctx.useTexCoords) {
> > -            for (auto& p : line) {
> > -                bbox.include(p.x, p.y);
> > -            }
> > +    uint16_t vertexDataOffset = _ctx.numVertices;
> > +
> > +    if (vertexDataOffset == 0) {
> > +        _ctx.indices = std::move(earcut.indices);
> > +    } else {
> > +        _ctx.indices.reserve(_ctx.indices.size() +  earcut.indices.size());
> > +        for (auto i : earcut.indices) {
>
> A reference would actually be larger than a uint16_t
>
> If we cared about optimizing this an insert may be faster or even a memcpy
> but it's a trivial amount of computation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tangrams/tangram-es/pull/268/files#r38588850>.
>



-- 
Hannes